### ISSUE TITLE
FP32 round-trip sum (fp16→fp32→sum→fp16) fails during compilation

### ISSUE THREAD

**Issue by mudhakar** [torch-spyre/torch-spyre#2619]
## Summary

Mixed-precision round-trip reduction — casting fp16 input to fp32, performing a stick-level sum, and casting back to fp16 — fails during compilation. The DDL template matching stage crashes because no template exists for fp32 stick-level reductions with a shared-layout pattern.

## Reproducer

```python
import torch
import torch_spyre  # noqa: F401

def test_roundtrip_sum():
    torch.manual_seed(42)
    x_cpu = torch.randn(4, 64, dtype=torch.float16)

    def fn(x):
        y = x.to(torch.float32)
        z = y.sum(dim=-1, keepdim=True)
        return z.to(x.dtype)

    expected = fn(x_cpu)

    x_dev = x_cpu.to(torch.device("spyre"))
    compiled_fn = torch.compile(fn, backend="inductor")
    actual = compiled_fn(x_dev).cpu()

    diff = (actual.float() - expected.float()).abs()
    assert diff.max().item() < 0.5, f"max_abs={diff.max().item()}"
```

## Failure

```
[DDC] start working on DSC: sum
terminate: map::at
```

## Analysis

The existing DDL templates cover fp16 stick reductions (`summeanmaxexx2.ddl`, shared layout with `scale_=-2`) and fp32 non-stick reductions (`sdsc_sumnonstick_fp32.json`, `scale_=-1`), but there is no template for **fp32 stick-level reductions** (shared layout, `scale_=-2`, `IEEE_FP32` / `stickSize=32`).

Additionally, the fp16 template's `operation_bind` with `[%type_fp16]` segfaults when encountering fp32 tensors rather than gracefully falling through.

## What's needed

1. A DDL template for fp32 shared-layout stick reductions (the fp32 analog of `summeanmaxexx2.ddl`'s sum path)
2. Graceful reject in existing fp16 templates when encountering fp32 tensors

## Branch

`debug-layernorm-roundtrip` on `mudhakar/torch-spyre` has the SDSC generation fixes for this path (5 commits, rebased on main).

**Comment by viji560** [comment]
@mudhakar 
The op that fails is  **fp32todl16** .  Needs updates to the front-end when generating `sdsc.json` for this op. The changes required are: 

- fill all dimensions in  `std::vector<PrimaryDimTypes> layoutDimOrder_` in `primaryDsInfo `and `AllocateNode`
- fill `std::vector<int> maxDimSizes_` in `AllocateNode`  for all dimensions with order matching  `layoutDimOrder_ `
- scale per dim (to represent reduction/broadcast): set `std::vector<double> scale_` to -2 with order matching `layoutDimOrder_` for input and output tensors, as the tensor was reduced in the previous operation.

Attached the edited `sdsc_2.json` for the op.  Verified that `dxp_standalone` passes with this. 

[sdsc_2_fixed.json](https://github.com/user-attachments/files/29042299/sdsc_2_fixed.json)

**Comment by moriohara** [comment]
@mudhakar https://github.com/torch-spyre/torch-spyre/pull/2927 should fix this.

**Comment by viji560** [comment]
@moriohara  -- Did you verify that [#2927](https://github.com/torch-spyre/torch-spyre/pull/2927)  generated the DSC as mentioned above?

**Comment by viji560** [comment]
@moriohara  -- were you able to verify that [https://github.com/torch-spyre/torch-spyre/pull/2927](2927) generated the DSC as mentioned above?

**Comment by manid2** [comment]
Hi @mudhakar, I tested this locally with the latest code across the full stack and can confirm there are no compilation errors.

```python
$ cat fp32/issue_2619.py
import torch
import torch_spyre  # noqa: F401

def test_roundtrip_sum():
    torch.manual_seed(42)
    x_cpu = torch.randn(4, 64, dtype=torch.float16)

    def fn(x):
        y = x.to(torch.float32)
        z = y.sum(dim=-1, keepdim=True)
        return z.to(x.dtype)

    expected = fn(x_cpu)

    x_dev = x_cpu.to(torch.device("spyre"))
    compiled_fn = torch.compile(fn, backend="inductor")
    actual = compiled_fn(x_dev).cpu()

    diff = (actual.float() - expected.float()).abs()
    assert diff.max().item() < 0.5, f"max_abs={diff.max().item()}"

test_roundtrip_sum()


$ python fp32/issue_2619.py
[Fast Path Debug] SUCCESS: 28 patches, 20 eval steps, 3 input symbols, 23 total symbols, output_size=4224
```

**Comment by moriohara** [comment]
@mudhakar @viji560 sorry for getting back to you late. As of today, the test case is passing, generating the following DSC. `layoutDimOrder_` and `maxDimSizes_` in `sdsc_2.json` are matching with those in the edited version while `scale_` isn't.  Perhaps this is because **fp32todl16** is not a reduction op. In fact, `sdsc_1.json` for **sum** has -2 in `scale_`

[sdsc_2.json](https://github.com/user-attachments/files/31270579/sdsc_2.json)

[sdsc_1.json](https://github.com/user-attachments/files/31270937/sdsc_1.json)

cc: @manid2

**Comment by moriohara** [comment]
@mudhakar @manid2 
Among shapes being used for reduction op tests from test_inductor_ops.py, the following shapes cause an error for this round-trip reduction pattern:
(64,), (5129,), (2,4,8,64)

https://github.com/torch-spyre/torch-spyre/blob/9fd980f3ac2c0c6e0df144a033cdca1687e250b6/tests/inductor/test_inductor_ops.py#L342-L348

**Comment by viji560** [comment]
@moriohara  -- thank you. Possible to share the sdsc for the failing tests?
